src/credentials/DeviceAttestationConstructor.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "DeviceAttestationConstructor.h"
+
+#include <lib/core/CHIPTLV.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <iostream>
+
+namespace chip {
+namespace Credentials {
+
+// context tag positions
+enum
+{
+    kCertificationDeclarationTagId = 1,
+    kAttestationNonceTagId         = 2,
+    kTimestampTagId                = 3,
+    kFirmwareInfoTagId             = 4,
+};
+
+CHIP_ERROR DeconstructAttestationElements(const ByteSpan & attestationElements, ByteSpan & certificationDeclaration,
```

Comment:
This method does not allow extraction of firmware info. Should be added as a TODO and left as a refactor.